src/Microsoft.ML.TimeSeries/SeasonalityDetector.cs

```diff
@@ -0,0 +1,307 @@
+        /// </summary>
+        /// <param name="r"></param>
+        /// <param name="bestFreq"></param>
+        /// <param name="secondFreq"></param>
+        /// <param name="timeSeriesLength"></param>
+        /// <returns></returns>
+        private static int FindActualPeriod(Complex[] r, int bestFreq, int secondFreq, int timeSeriesLength)
+        {
+            int firstPeriod = -1;
+            int secondPeriod = -1;
+            double firstTimeDomainEnergy = -1;
+            double secondTimeDomainEnergy = -1;
+            firstPeriod = FindBestPeriod(r, bestFreq, timeSeriesLength, out firstTimeDomainEnergy);
+            if (secondFreq != -1)
+            {
+                secondPeriod = FindBestPeriod(r, secondFreq, timeSeriesLength, out secondTimeDomainEnergy);
+            }
+            if (firstPeriod == -1 && secondPeriod == -1)
+                return -1;
+            int truePeriod;
+            double trueTimeDomainEnergy;
+            if (firstPeriod == -1)
+            {
+                truePeriod = secondPeriod;
+                trueTimeDomainEnergy = secondTimeDomainEnergy;
+            }
+            else if (secondPeriod == -1)
+            {
+                truePeriod = firstPeriod;
+                trueTimeDomainEnergy = firstTimeDomainEnergy;
+            }
+            else
+            {
+                if (firstPeriod == secondPeriod)
+                {
+                    truePeriod = firstPeriod;
+                    trueTimeDomainEnergy = firstTimeDomainEnergy;
+                }
+                else
+                {
+                    // hueristic: if the second frequency is with somewhat higher energy in time domain, we think it is a better candidate
+                    if (secondTimeDomainEnergy > firstTimeDomainEnergy * 1.05)
+                    {
+                        truePeriod = secondPeriod;
+                        trueTimeDomainEnergy = secondTimeDomainEnergy;
+                    }
+                    else
+                    {
+                        truePeriod = firstPeriod;
+                        trueTimeDomainEnergy = firstTimeDomainEnergy;
+                    }
+                }
+            }
+            trueTimeDomainEnergy /= r[0].Real;
+
+            /* This is a key equation, which is named the "testing for randomness with the correlogram". /ref: http://www.ltrr.arizona.edu/~dmeko/notes_3.pdf
+             * 1.96 is for the 2-sigma, which has 95% statistical confidence. 2.58 is for 99% confidence, 2.85 for 99.5% confidence
+             * increasing the threshold aims to mitigate the fake seasonal component caused by outliers. in practice, if there exist true seasonal component,
+             * such as BirdStrike/Appdownloads, the energy is far larger than threshold, hence change threshold from 2.85 to 4.0 have no impact (tested);
+             */
+            double threshold = 4 / Math.Sqrt(timeSeriesLength);
```

Comment:
Should we not expose confidence as a parameter?

Reply:
done